lib/edge_impulse/ei_ncs.cpp

```diff
@@ -0,0 +1,399 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <assert.h>
+#include <ei_run_classifier.h>
+#include <ei_ncs.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(ei_ncs, CONFIG_EI_NCS_LOG_LEVEL);
+
+#define INPUT_FRAME_SIZE	EI_CLASSIFIER_RAW_SAMPLES_PER_FRAME
+#define INPUT_WINDOW_SIZE	EI_CLASSIFIER_DSP_INPUT_FRAME_SIZE
+#define HAS_ANOMALY		EI_CLASSIFIER_HAS_ANOMALY
+
+BUILD_ASSERT(CONFIG_EI_NCS_THREAD_STACK_SIZE > 0);
+
+#define DATA_BUFFER_SIZE	CONFIG_EI_NCS_DATA_BUF_SIZE
+#define THREAD_STACK_SIZE	CONFIG_EI_NCS_THREAD_STACK_SIZE
+#define THREAD_PRIORITY 	CONFIG_EI_NCS_THREAD_PRIORITY
+#define DEBUG_MODE		IS_ENABLED(CONFIG_EI_NCS_DEBUG_MODE)
+
+enum state {
+	STATE_DISABLED,
+	STATE_PROCESSING,
+	STATE_READY,
+};
+
+struct data_buffer {
+	float buf[DATA_BUFFER_SIZE];
+	float *processing_head;
+	float *append_head;
+	size_t wait_data_size;
+	struct k_spinlock lock;
+	enum state state;
+};
+
+static K_THREAD_STACK_DEFINE(thread_stack, THREAD_STACK_SIZE);
+static struct k_thread thread;
+static k_tid_t ei_thread_id;
+
+static K_SEM_DEFINE(ei_sem, 0, 1);
+
+static struct data_buffer ei_input;
+static ei_impulse_result_t ei_result;
+static ei_ncs_result_ready_cb user_cb;
+
+
+BUILD_ASSERT(DATA_BUFFER_SIZE > INPUT_WINDOW_SIZE);
+BUILD_ASSERT(INPUT_WINDOW_SIZE % INPUT_FRAME_SIZE == 0);
+
+
+static size_t buf_calc_max_processing_move(const struct data_buffer *b)
+{
+	const float *buf_start = b->buf;
+	const float *buf_end = buf_start + ARRAY_SIZE(b->buf);
+
+	if (b->append_head >= b->processing_head) {
+		return b->append_head - b->processing_head;
+	}
+
+	return (buf_end - b->processing_head) + (b->append_head - buf_start);
+}
+
+static size_t buf_calc_free_space(const struct data_buffer *b)
+{
+	const float *buf_start = b->buf;
+	const float *buf_end = buf_start + ARRAY_SIZE(b->buf);
+
+	if (b->wait_data_size > 0) {
+		return b->wait_data_size + ARRAY_SIZE(b->buf) -
+		       INPUT_WINDOW_SIZE - 1;
```

Comment:
Why you subtract `INPUT_WINDOW_SIZE ` in this case? Other cases does not leave space for window.

Reply:
What if processing is not active? Is this function not used then?

Reply:
Maybe it could be:
`get_busy_data_count() {return right_value;}` and then
`get_free_data_space() {return b->wait_data_count + buf_size - get_busy_data_count() - 1;}`

Reply:
wait_data_size can have any value - it can be even bigger than the array, so I do not use the pointer/index logic here. After we finish waiting for the input data, we will have whole array of free space except the input window that will currently be processed.

Reply:
This function can be used no matter if processing is active or not.

Reply:
I will work on simplifying this one when I will change pointers to indexes.

Reply:
discussed offline - you look at the other end; that's fine